### PR TITLE
Throw on data dependent outputs

### PR DESCRIPTION
**Pull request by eellison**
We need to flip on https://github.com/pytorch/pytorch/pull/83567 for correctness. This patch allows us to continue to trace through `.item()` by catching the exception.

It also switches the optimizers to use `fake_tensors=False` because as explained in https://github.com/pytorch/pytorch/issues/84597 and https://github.com/pytorch/torchdynamo/issues/1138, the current wrapping of `.item()` for those cases only defers the `_local_scaler_dense` call until torch dispatch layer, where fakes tensors have to throw. Those issues lay out a path forward for not graph breaking, but we should get this patch in even with the graph breaks because it's needed for correctness. 

This is ready for review but it's going to take another day before the changes in pytorch are available in the nightly and the tests will pass.

### CODE REVIEW

**Review by eellison — COMMENTED** [review]
This is being tested here https://github.com/pytorch/torchdynamo/blob/main/test/test_repros.py#L887

File: torchdynamo/variables/tensor.py [hunk 1]
```diff
@@ -145,32 +148,26 @@ def context():
...
+                    if use_fake_tensors and isinstance(e, DataDependentOutputException):
+                        if (
+                            config.capture_scalar_outputs
+                            and proxy.node.target == "item"
+                        ):
+                            example_value = torch.zeros(
+                                size=(), dtype=args[0].dtype
+                            ).item()
+                        else:
+                            unimplemented("data dependent operator")
```
Comment (voznesenskym) [inline torchdynamo/variables/tensor.py:160]: nit - can you add that it's .item here? either hardcoded or just via `f" ... {proxy.node.target}"` - makes it easier to find and debug, also makes the torchdynamo.explain graph breaks easier to grok.

File: torchdynamo/variables/tensor.py [hunk 2]
```diff
@@ -145,32 +148,26 @@ def context():
...
+                    elif use_fake_tensors and isinstance(
+                        e, DynamicOutputShapeException
+                    ):
+                        unimplemented("dynamic shape operator")
```
Comment (voznesenskym) [inline torchdynamo/variables/tensor.py:164]: Is there any data from the exception we can plumb here for a nicer message?

Comment (eellison) [reply torchdynamo/variables/tensor.py:164]: we can print the dynamic shape operator, yea

File: torchdynamo/variables/tensor.py [hunk 3]
```diff
@@ -145,32 +148,26 @@ def context():
...
                         )
                 except RuntimeError as e:
-                    raise TorchRuntimeError() from e
+                    if use_fake_tensors and isinstance(e, DataDependentOutputException):
+                        if (
+                            config.capture_scalar_outputs
+                            and proxy.node.target == "item"
+                        ):
+                            example_value = torch.zeros(
+                                size=(), dtype=args[0].dtype
```
Comment (voznesenskym) [inline torchdynamo/variables/tensor.py:157]: the args here could be malformed, can you assert that:

args len is 1
args[0] is a tensor
args[0] is a numel() of 1

Comment (eellison) [reply torchdynamo/variables/tensor.py:157]: This is post-dispatch, so it should already have thrown in that case

Comment (voznesenskym) [reply torchdynamo/variables/tensor.py:157]: Ah yea sg

File: test/test_optimizers.py [hunk 4]
```diff
@@ -71,16 +71,19 @@ def setUpClass(cls):
...
 
     # ASGD has a small optimization that avoids averaging
     # This will fully capture the graph once that optimization is removed
     # NB: in python versions < 3.8, we don't capture graphs when breaks
     # occur in a loop
-    test_asgd = make_test(
-        torch.optim.ASGD, exp_frame_cnt=(0 if sys.version_info < (3, 8) else 6)
-    )
+
+    # Fails without fake tensor:
```
Comment (voznesenskym) [inline test/test_optimizers.py:82]: Interesting. Can you file an issue for this? Cc @mlazos

**Review by voznesenskym — APPROVED** [review]
Seems reasonable

File: test/test_unspec.py [hunk 5]
```diff
@@ -133,6 +133,7 @@ def fn(x):
         res2 = opt_fn(x)
         self.assertTrue(same(res1, res2))
 
+    @patch.object(torchdynamo.config, "fake_tensor_propagation", False)
```
Comment (yanboliang) [inline test/test_unspec.py:136]: I'm curious why this failed? This is one of the test case copied from Pytorch CI, where fake tensor is enabled IIRC.

Comment (eellison) [reply test/test_unspec.py:136]:
```
>>> x = torch.tensor(0)
>>> torch.randint(x, x, x)
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
TypeError: randint() received an invalid combination of arguments - got (Tensor, Tensor, Tensor), but expected one of:
 * (int high, tuple of ints size, *, torch.Generator generator, Tensor out, torch.dtype dtype, torch.layout layout, torch.device device, bool pin_memory, bool requires_grad)
 * (int high, tuple of ints size, *, Tensor out, torch.dtype dtype, torch.layout layout, torch.device device, bool pin_memory, bool requires_grad)
 * (int low, int high, tuple of ints size, *, torch.Generator generator, Tensor out, torch.dtype dtype, torch.layout layout, torch.device device, bool pin_memory, bool requires_grad)
 * (int low, int high, tuple of ints size, *, Tensor out, torch.dtype dtype, torch.layout layout, torch.device device, bool pin_memory, bool requires_grad)
 ```
 
 the int->torch.tensor(0) conversion causes an error

**Pull request merged**